api/scripts/daily-digest.ts

```diff
@@ -0,0 +1,125 @@
+import type { MonitoringParams } from "../api/V1/monitoringPage.js";
+import { calculateMonitoringPageCriteria } from "../api/V1/monitoringPage.js";
+import { generateVisits } from "../api/V1/monitoringVisit.js";
+import modelsInit from "@models/index.js";
+import { RecordingType } from "@typedefs/api/consts.js";
+
+(async () => {
+  try {
+    const models = await modelsInit();
+    const users = await models.User.findAll({where: {
+      'settings.emailNotifications.dailyDigest': true
+    }});
+    
+    const templateFilename = "daily-digest.html";
+    const client = new SMTPClient(config.smtpDetails);
+    // const client = new SMTPClient({
+    //   user: "",
+    //   password: "",
+    //   host: "smtp.gmail.com",
+    //   ssl: true,
+    // });
+
+    users.forEach(async user => {
+      const requestUser = user;
+      const groups = await models.GroupUsers.findAll({where: {
+        'UserId': user.id
+      }});
+
+      groups.forEach(async group => {
+        const recordingData = {};
+        const name = await models.Group.findOne({where: {
+          'id': group.GroupId
+        }});
+
+        const stations = await models.Station.findAll({where: {
+          'GroupId': group.GroupId
+        }});
+        const stationsArray = [];
+        stations.forEach(async station => {
+          stationsArray.push(station.id);
+        });
+
+        const oneWeekAgo = new Date();
+        const now = new Date();
+        oneWeekAgo.setHours(oneWeekAgo.getHours() - 168);
+
+        const params: MonitoringParams = {
+          stations: stationsArray,
+          groups: [group.GroupId],
+          page: 1,
+          pageSize: 100,
+          from: oneWeekAgo,
+          until: now,
+          types: [RecordingType.ThermalRaw]
+        };
+
+        const viewAsSuperAdmin = true;
+        const searchDetails = await calculateMonitoringPageCriteria(
+          requestUser,
+          params,
+          viewAsSuperAdmin
+        );
+      
+        searchDetails.compareAi = "Master";
+        searchDetails.types = params.types;
+      
+        const visits = await generateVisits(
+          requestUser.id,
+          searchDetails,
+          viewAsSuperAdmin
+        );
+        
+        (visits as unknown as Array<any>).forEach(async visit => {
+          if (recordingData[visit.classification] !== undefined) {
+            recordingData[visit.classification] += 1;
+          } else {
+            recordingData[visit.classification] = 1;
+          }
+        });
+
+        const visitsTotal = (visits as unknown as Array<any>).length;
+        const speciesListArray = Object.entries(recordingData).map(([species, count]) => {
+          return { species, count, widthPercent: 100 / Object.keys(recordingData).length };
+        });
+
+        const interpolants: {
+          [key: string]: any;
+        } = {
+          groupName: `${name.groupName}`,
+          groupURL: `https://browse-next.cacophony.org.nz/${name.groupName}`,
```

Comment:
Group/Project names can be anything, and they're not necessarily URL safe:  There's a function `api/emails/htmlEmailUtils.ts:urlNormaliseName` to deal with this.